arangod/RestHandler/RestDocumentHandler.cpp

```diff
@@ -193,6 +194,12 @@ RestStatus RestDocumentHandler::insertDocument() {
     return RestStatus::DONE;
   }
 
+  if (ServerState::instance()->isDBServer() &&
+      _activeTrx->state()->collection(cname, AccessMode::Type::WRITE) == nullptr) {
+    generateTransactionError(cname, OperationResult(Result(TRI_ERROR_CLUSTER_SHARD_FOLLOWER_REFUSES_OPERATION), opOptions), "");
+    return RestStatus::DONE;
+  }
```

Comment:
It is possible that this happens on a follower.
The leader will at some point, potentially lazily, send a transaction begin request to a follower as part of a replicated write operation. Included in this transaction begin request are the affected (written-to) shards. In edge cases it is possible that the leader puts together the list of followers for a shard for an operation, and then (a bit later) creates the actual requests to these followers. If in the meantime a follower got dropped by a concurrent operation, there is a small window of time in which a request is generated for an already-dropped follower. In this case, the `buildTransactionBody()` function in ClusterTrxMethods.cpp will not include the target shard in the transaction begin request. 
If such requests arrive on followers, they previously failed with assertion failures because it was attempted to write in a read-only transaction (read-only because no shards were listed in either "write" or "exclusive"). With the change, the follower will instantly reject such requests.

Reply:
Okay, so this code (in this case) is executed on the follower, correct? So in case we got such a request like you described here, we're executing

```cpp
...
  _activeTrx = createTransaction(cname, AccessMode::Type::WRITE, opOptions);
...
  Result res = _activeTrx->begin();
// res.ok()
...
  _activeTrx->state()->collection(cname, AccessMode::Type::WRITE) == nullptr
```

. But if this follower got dropped for that shard, I still don't quite understand how `createTransaction` (and `begin()`) can succeed at all in that case?

Reply:
The follower got dropped for the shard, but it does not need to know about it when we are here. It only gets a request to begin a transaction, however, the request does not include its own shard in either "write" or "exclusive". The whole point of this extra check is to detect this, and make the operation fail early.
If we don't do this, it may believe that it is still a follower further inside, and try to carry out the write operation inside the read-only transaction, which will lead to an assertion failure

Reply:
Okay thanks.

I think though that the check you added should be moved inside `RestVocbaseBaseHandler::createTransaction`, so that we no longer can successfully order a transaction for a collection that does not contain this collection. This will prevent something similar from happening in other places as well.

Reply:
As far as I see `RestVocbaseBaseHandler::createTransaction` handles three cases:
1. No transaction ID was given as HTTP header: then a new SingleCollectionTransaction is created for the given collection.
2. A trxId was given and that transaction already exists.
3. A trxId was given and no transaction exists and is created lazily here.
In case 1. we do not need to check that the newly created transaction contains the given collection.
Cases 2. and 3. are treated the same here, since we are using `ensureManagedTrx` in these cases we can add the check that the transaction has the given collection with the given write mode, and can throw an exception if this is not the case.
I will propose a patch.